src/Datadog.Trace/Util/Clock.cs

```diff
@@ -0,0 +1,54 @@
+using System;
+
+namespace Datadog.Trace.Util
+{
+    internal class Clock
+    {
+        /// <summary>
+        /// Flag used to avoid checking the threadstatic field when running outside of tests
+        /// </summary>
+        private static bool _overrideClock;
+
+        [ThreadStatic]
+        private static IClock _customClock;
+
+        public static DateTime UtcNow
+        {
+            get
+            {
+                if (!_overrideClock)
+                {
+                    return DateTime.UtcNow;
+                }
+
+                return _customClock?.UtcNow ?? DateTime.UtcNow;
+            }
+        }
+
+        internal static void Reset()
+        {
+            _overrideClock = false;
+        }
+
+        /// <summary>
+        /// Overrides the clock used by the current thread.
+        /// This method should be called only from unit tests.
+        /// </summary>
+        /// <param name="customClock">Fake clock</param>
+        /// <returns>Lease to dispose to restore the original state</returns>
+        internal static IDisposable SetForCurrentThread(IClock customClock)
+        {
+            _overrideClock = true;
+            _customClock = customClock;
+            return new Lease();
+        }
+
+        private class Lease : IDisposable
+        {
+            public void Dispose()
+            {
+                _customClock = null;
```

Comment:
Should we set `_overrideClock = false` in `Lease.Dispose()`, or did you leave that out intentionally? I see it's reset in `Clock.Reset()` instead.

Do we need one way to set `_customClock = null` and another for `_overrideClock = false`? In what scenario would we use one and not the other?

Do we need `_overrideClock` at all? Isn't it enough to check if `_customClock` is null or not?

Reply:
The more of these static types we try to mock, the more I think we should inject them as services, DI-style. For example, we could pass `IClock` into the `RateLimiter` ctor and `IIdGenerator` into `SpanContext`.

Reply:
Reading a threadstatic field has a cost, and I feel really bad about adding that cost to the app just for testability purpose 😛  the `_overrideClock` flag is there to skip the TLS field when running outside of tests. It's just a few nanoseconds though, so I can remove `_overrideClock` and put it back in the unlikely case it shows up as a hotspot in the future. As you wish.

I'm usually a huge fan of DI, but I don't if it translates well to our code, since the API itself is static and relies on ambient state. If we try it, I think it should be in a separate PR?

Reply:
> I feel really bad about adding that cost to the app just for testability purpose

Agreed. Let's keep the two fields if it helps avoid the extra cost for users.

Regarding DI: Yes I meant that as something to think about in the future, definitely not in this PR 😉